SSA/Projects/InstCombine/AliveAutoGenerated.lean

```diff
@@ -58,6 +59,78 @@ theorem alive_AddSub_1043  (w : Nat)   : alive_AddSub_1043_src w  ⊑ alive_AddS
   simp_alive_peephole
   apply bitvec_AddSub_1043
 
+-- Name: SimplifyDivRemOfSelect
+-- precondition: true
+/-
+%sel = select %c, %Y, 0
+%r = udiv %X, %sel
+  =>
+
+%r = udiv %X, %Y
+
+-/
+def alive_DivRemOfSelect_src (w : Nat) :=
+[mlir_icom (w)| {
+^bb0(%c: i1, %y : _, %x : _):
+  %c0 = "llvm.mlir.constant" () { value = 0 : _ } :() -> (_)
+  %v1 = "llvm.select" (%c,%y, %c0) : (i1, _, _) -> (_)
+  %v2 = "llvm.udiv"(%x, %v1) : (_, _) -> (_)
+  "llvm.return" (%v2) : (_) -> ()
+}]
+
+def alive_DivRemOfSelect_tgt (w : Nat) :=
+[mlir_icom (w)| {
+^bb0(%c: i1, %y : _, %x : _):
+  %v1 = "llvm.udiv" (%x,%y) : (_, _) -> (_)
+  "llvm.return" (%v1) : (_) -> ()
+}]
+
+#check alive_DivRemOfSelect_src
+#check alive_DivRemOfSelect_tgt
+
+@[simp]
+theorem BitVec.ofNat_toNat_zero : 
+BitVec.toNat (BitVec.ofInt w 0) = 0 := 
+  by simp[BitVec.toNat, BitVec.ofInt, BitVec.toFin, BitVec.ofNat, OfNat.ofNat]
+
+theorem alive_DivRemOfSelect (w : Nat) : 
+  alive_DivRemOfSelect_src w ⊑ alive_DivRemOfSelect_tgt w := by
+ unfold alive_DivRemOfSelect_src alive_DivRemOfSelect_tgt
+ intros Γv
+ simp_peephole at Γv
+ simp (config := {decide := false}) only [OpDenote.denote,
+   InstCombine.Op.denote, HVector.toPair, HVector.toTriple, pairMapM, BitVec.Refinement,
+   bind, Option.bind, pure, DerivedContext.ofContext, DerivedContext.snoc,
+   Ctxt.snoc, MOp.instantiateCom, InstCombine.MTy.instantiate,
+   ConcreteOrMVar.instantiate, Vector.get, HVector.toSingle,
+   LLVM.and?, LLVM.or?, LLVM.xor?, LLVM.add?, LLVM.sub?,
+   LLVM.mul?, LLVM.udiv?, LLVM.sdiv?, LLVM.urem?, LLVM.srem?,
+   LLVM.sshr, LLVM.lshr?, LLVM.ashr?, LLVM.shl?, LLVM.select?,
+   LLVM.const?, LLVM.icmp?,
+   HVector.toTuple, List.nthLe, bitvec_minus_one]
+ intros y
+ intros x
+ intros c
+ simp[Option.bind, bind, Monad.toBind]
+ clear Γv
+ cases c
+ -- | select condition is itself `none`, nothing more to be done. propagate the `none`.
+ . case none => cases x <;> cases y <;> simp
+ . case some cond =>
+    obtain ⟨vcond, hcond⟩ := cond
+    have hcond' : vcond = 1 ∨ vcond = 0 := by 
+      norm_num at hcond
+      rcases vcond with zero | vcond <;> simp <;>
+      rcases vcond with zero | vcond <;> simp <;>
+      linarith
+    cases hcond'
+    . case inl h => -- have selected 0, so we compute X/0 which is UB.
+      subst h
+      simp
+    . case inr h => -- have selected Y itself, so we compute X/Y.
+      subst h; simp
+      cases' x with vx <;> 
+      cases' y with vy <;> simp
```

Comment:
This certainly does not belong in an auto-generated file, no?

Reply:
It is unclear where to keep this. The proof state after simp_alive_peephole is not really usable.

We cannot directly have the full theorem statement before elaboration since we do not import the EDSL into AliveStatements.


After elaboration but before simp_alive_peephole, the theorem statement is quite large, and so I chose not to copy that into AliveStatements.lean.

Reply:
If this cannot yet be auto-generated, maybe we should keep it in a separate file, e.g., AliveManual.lean? The auto-generated file is supposed to only contain auto-generated content. We added trivial references to external proofs as they started to work, as these were easy to add and could eventually also be auto-generated. Adding large amounts of manual stuff will prevent us from regenerating this file.